src/renderer.cpp

```diff
@@ -596,25 +625,173 @@ void Renderer::shutdown()
+
+    // Set current view transform
+    bgfx::setViewTransform(backendViewId, internalState.viewMatrix, internalState.backendProjMatrix);
+
+    // Create a world matrix
+    struct matrix worldMatrix;
+    identity_matrix(&worldMatrix);
+
+    // WalkMesh offset to adjust vertical positon of walkmesh so that it is just under the character feets
+    worldMatrix._43 = walkMeshPosOffset;
 
-        bgfx::setViewRect(backendViewId, 0, 0, framebufferWidth, framebufferHeight);
+    struct matrix viewMatrix;
+    memcpy(&viewMatrix.m[0][0], internalState.viewMatrix, sizeof(float) * 16);
 
-        if (internalState.bDoScissorTest) bgfx::setScissor(scissorOffsetX, scissorOffsetY, scissorWidth, scissorHeight);
+    struct matrix worldViewMatrix;
+    multiply_matrix(&worldMatrix, &viewMatrix, &worldViewMatrix);
+
+    memcpy(internalState.worldViewMatrix, &worldViewMatrix.m[0][0], sizeof(float) * 16);
+    internalState.bIsTLVertex = false;
+    internalState.blendMode = RendererBlendMode::BLEND_NONE;
+    internalState.bIsFBTexture = false;
+    internalState.bIsTexture = false;
+
+    // Set draw type to field shadow mode
+    setDrawType(DrawType::FIELD_SHADOW);
+
+    setCommonUniforms();
+
+    // Bind shadow map with comparison sampler
+    bgfx::setTexture(3, getUniform(shaderTextureBindings[3], bgfx::UniformType::Sampler), bgfx::getTexture(shadowMapFrameBuffer));
+
+    // Bind shadow map for direct depth sampling
+    uint32_t flags = BGFX_TEXTURE_RT | BGFX_SAMPLER_POINT | BGFX_SAMPLER_U_CLAMP | BGFX_SAMPLER_V_CLAMP | BGFX_SAMPLER_W_CLAMP;
+    bgfx::setTexture(4, getUniform(shaderTextureBindings[4], bgfx::UniformType::Sampler), bgfx::getTexture(shadowMapFrameBuffer), flags);
+
+    // Set state
+    {
+        internalState.state = BGFX_STATE_WRITE_RGB | BGFX_STATE_WRITE_A;
+
+        internalState.state |= BGFX_STATE_CULL_CCW;
+
+        internalState.state |= BGFX_STATE_BLEND_EQUATION(BGFX_STATE_BLEND_EQUATION_ADD);
+        internalState.state |= BGFX_STATE_BLEND_FUNC(BGFX_STATE_BLEND_SRC_ALPHA, BGFX_STATE_BLEND_INV_SRC_ALPHA);
+
+        internalState.state |= BGFX_STATE_DEPTH_TEST_LEQUAL;
+        internalState.state |= BGFX_STATE_WRITE_Z;
     }
+    bgfx::setState(internalState.state);
+
+    uint32_t stencilState = BGFX_STENCIL_TEST_EQUAL;
+    stencilState |= BGFX_STENCIL_FUNC_REF(255);
+    stencilState |= BGFX_STENCIL_FUNC_RMASK(0xff);
+    stencilState |= BGFX_STENCIL_OP_FAIL_S_KEEP;
+    stencilState |= BGFX_STENCIL_OP_FAIL_Z_KEEP;
+    stencilState |= BGFX_STENCIL_OP_PASS_Z_KEEP;
+    bgfx::setStencil(stencilState, stencilState);
+
+    bgfx::submit(backendViewId, backendProgramHandles[backendProgram]);
+}
+
+void Renderer::draw()
+{
+    if (enable_lighting && internalState.isCastShadow)
+    {
+        // Save current view id
+        int tmpBackendViewId = backendViewId;
+
+        backendViewId = 0;
+        drawShadowMap();

```

Comment:
If you need to ensure that your code runs before the current draw call, just fire `drawShadowMap` as the first line after
```cpp
::draw()
{
 // your code here
 
 // rest of the code here
}
```
without mangling the current `backendViewId`. Or am I missing something?

Reply:
That would not work because I need the whole scene to be drawn to the shadowmap before any normal rendering is done, not just the model currently being rendered.
If I just called `drawShadowMap` without changing the view Id to a lower value (0 for the shadowmap), only the currently model being drawn and previously drawn models would exist in the shadow map referenced during normal rendering.
For example if there is box placed on the floor, and the floor is drawn first, the box would not have been drawn to the shadowmap yet and therefore no shadow would be cast on the floor. 
Hope this clears out the reason why there is a need to reserve view id 0 for the shadowmap rendering.

Reply:
Aha, ok then you can consider doing that in this in the method: https://github.com/julianxhokaxhiu/FFNx/blob/master/src/renderer.cpp#L317

Because it's the best moment as you get the entire scene, drawn already and before showing it on the Backbuffer, you can do all the magic you need to ( basically it's also where the "post" effects are done, which atm are a simple plain rendering of the Framebuffer in the Backbuffer ).

Wouldn't this work? Also this would guarantee you have your own entire attachments cleared and so on.

Reply:
No, that won't work because as I explained before, drawing to the shadowmap is a pre-process that needs to be done _**before**_ normal rendering, not as a post-process. The renderFrame method is called _**after**_ normal rendering and as you said it's where post effects are done.

I don't know if you are familiar with the shadowmap algorithm but it is not a post-process. To be fair, field shadows on the background are implemented somehow like a post-process where shadows are drawn by darkening the 2D field background. However, this is a physically incorrect way of doing shadows, but because field backgrounds are just 2D, the only way is to fake it as a post-process. 

On the other hand, characters and the battle background are all 3D models where accurate lighting and shadows can be achieved. Here, shadows are not a post-process but are calculated during normal rendering by fetching the pre-computed shadowmap and using it as an occlusion term to the diffuse and specular components of the luminance calculation:
[https://github.com/julianxhokaxhiu/FFNx/blob/427166ede3b2b33dc5a8c959c63fb9d48d2bc212/misc/FFNx.lighting.sh#L100](url)

Please notice that the shadow factor only applies to the diffuse and specular terms and _**not**_ the ambient term. This calculation cannot be achieved as a post-process in a forward renderer such as the one used in FFNx. By the way, shadow mapping as explained above is nothing unusual, it is how shadows are implemented in most games.

As explained above, the shadowmap rendering is a pre-process step and therefore requires a view id  with a value lower than the normal rendering view id. It is not possible to just use the same view id as normal rendering and call the submit function for shadowmap rendering before normal rendering. This is because the following reasons:
- The shadowmap rendering uses a different framebuffer where the scene is drawn from the light point of view, not the game camera point of view
- The shadowmap needs to be referenced as a texture during normal rendering

Therefore, I think that reserving view id 0 for the shadowmap rendering is the most reasonable way of doing it. As explained before, the only change now is that the 'backendViewId' variable now starts at 1 but that should not change the behavior of the renderer in any way.

Reply:
Allright, I understand the whole point and now it starts to make sense.

So, if that's the case I suggest then in order to tackle this PR definitely:
1. Consider moving back the shadow vert and frag logics into separate shader files ( sorry! )
2. Fine by reserving the view id 0, just make sure in the code the shaders attached to that view are only the one required for the shadowmap calculation
3. Once all of the above is done, PLEASE consider testing your own code on FF8 and make sure you didn't break anything

Finally, as you have shielded all the light code into a flag, please make sure the shadowmap processing logics ( methods, code, etc. ) are as isolated as possible from the main code ( which means do not change AT ALL the `draw()` method code, just add your own method call where you need to ). This will make the code also much more easily to be managed.

Additionally please do not bind the whole uniform set of variables if those are required only by the Shadowmap, make sure those are attached only to your own shaders ( which means do not change ANYTHING from the existing core and post frag and vert files ).

Let's see once this whole bunch of changes are done how the code looks like and if it's good, the only part I'll test is how it works on FF8 just to make sure we didn't break anything.

Just to be clear, what I expect after this new set of changes are:
1) Minimal changes to the existing code ( by minimal I mean adding your function calls where needed, but the structure is not changed, not like how the `renderer.cpp` looks like now, or the `common.cpp` too
2) A ton of new files with their own logics
3) A complete new set of lines in the `renderer.cpp` which are your own new methods and members required to accomplish the goal
4) MINIMAL glue code, shielded by the flag and by a check which ensure it's running only on ff7

Looking forward into this :) Sorry for the challange, but this last comment made me recognize the shape of this code and how I expect it to be now.

Reply:
Thank you for understanding the need to reserve view id 0 for shadowmap rendering.
> Consider moving back the shadow vert and frag logics into separate shader files ( sorry! )

May I ask what made you change your mind on this? Also does this refer only to the shadowmap rendering shader or also the shader that rendered the walkmesh for the field shadows?

> Once all of the above is done, PLEASE consider testing your own code on FF8 and make sure you didn't break anything

As I said in a previous comment, FF8 is region blocked and not available for purchase/activation in my country, meaning that it is impossible for me to test if anything is broken in FF8. I would appreciate if you can test it for me, but if you are not available then maybe I can ask for help to someone else on the qhimm community.

> Finally, as you have shielded all the light code into a flag, please make sure the shadowmap processing logics ( methods, code, etc. ) are as isolated as possible from the main code ( which means do not change AT ALL the draw() method code, just add your own method call where you need to ). This will make the code also much more easily to be managed.

If I understand correctly, this means reverting to a similar state as when I first submitted this PR, where a separate drawShadowMap() function was used, is that correct? Also may I ask why this change of mind?

Also you say to not change anything in the draw() function but I need the following two changes for it to work:
1. Bind the shadowmap texture. This could be done without touching the draw function with the Renderer::useTexture() function, is that correct?
2. Set the stencil buffer. There is not currently a way to set the stencil buffer so I was thinking to add some APIs to set the stencil function and reference value from outside the draw() function. However, the actual setting by calling the bgfx functions would be inside the draw() function in similar to how the blend state is set. Would that be ok?

> Additionally please do not bind the whole uniform set of variables if those are required only by the Shadowmap, make sure those are attached only to your own shaders ( which means do not change ANYTHING from the existing core and post frag and vert files ).

You say here to not change anything from the core shaders, but lighting calculations as well as fetching the shadowmap needs to be done here so I'm not sure I understand your request (most of it is already is in include files anyway). 

> Minimal changes to the existing code ( by minimal I mean adding your function calls where needed, but the structure is not changed, not like how the renderer.cpp looks like now, or the common.cpp too

If I understand you correctly, this means reverting to something similar as when I was using separate draw functions such as drawShadowMap() but instead of implementing them inside the Renderer class, I create a different class and source files (lighting.h/cpp) that contains the lighting related draw functions as well as the code found in common.cpp. Is this correct?

> A ton of new files with their own logics

Other than the lighting class files described above, I was thinking to separate the walkmesh extraction/extrusion functions into separate files (currently in field.h/cpp). Other than that I don't see what else would require new files. I would appreciate if you specify what else you want me to separate.

Thank you again for your time reviewing this PR, it is very much appreciated!

Reply:
> Thank you for understanding the need to reserve view id 0 for shadowmap rendering.

Thank you for taking the patience for still working on this. Really, really appreciate it!

> May I ask what made you change your mind on this? Also does this refer only to the shadowmap rendering shader or also the shader that rendered the walkmesh for the field shadows?

As we do reserve an entire view, it makes sense to use also separate shaders. As you will probably fire your own draw calls or whatever needed, you can totally bind your own shaders when calling them, and leave the game shaders to draw as usual. This makes this PR safer to merge as "if I enable the flag, I get something done in the view 0, if not, I still get what I used to have, but starting from view 1". That's basically a way to "kill" and entire block of code with a piece of mind when merging :)

> As I said in a previous comment, FF8 is region blocked and not available for purchase/activation in my country, meaning that it is impossible for me to test if anything is broken in FF8. I would appreciate if you can test it for me, but if you are not available then maybe I can ask for help to someone else on the qhimm community.

Not sure where are you living at the moment, but you can fetch a copy from archive.org. Just saying :D

> If I understand correctly, this means reverting to a similar state as when I first submitted this PR, where a separate drawShadowMap() function was used, is that correct? Also may I ask why this change of mind?

See the answer above. That applies also for this case. Related to the previous state, yes, feel free to encapsulate all your code in a separate function, or better, see the answers below :)

> Bind the shadowmap texture. This could be done without touching the draw function with the Renderer::useTexture() function, is that correct?

No, that function is to bind texture attachments when drawing something. Not sure how the light texture works but if that makes your day, feel free to use it ( as it is! ) in your own function code :)

> Set the stencil buffer. There is not currently a way to set the stencil buffer so I was thinking to add some APIs to set the stencil function and reference value from outside the draw() function. However, the actual setting by calling the bgfx functions would be inside the draw() function in similar to how the blend state is set. Would that be ok?

Absolutely fine to have the Stencil part exactly like we have the blend state for eg. Feel free to implement an enum, a relative function to set which state you want and let the `draw()` API set it based on which is the current enum set. Remember to reset its state in the `reset()` function. That's fine by me!

> You say here to not change anything from the core shaders, but lighting calculations as well as fetching the shadowmap needs to be done here so I'm not sure I understand your request (most of it is already is in include files anyway).

Whatever calculation you need to do, consider doing it in your own new shader vert and frag files. Use those ones as your own greenfield and remember to attach them when drawing :)

> If I understand you correctly, this means reverting to something similar as when I was using separate draw functions such as drawShadowMap() but instead of implementing them inside the Renderer class, I create a different class and source files (lighting.h/cpp) that contains the lighting related draw functions as well as the code found in common.cpp. Is this correct?

Well, that's up to you if it makes it easier. For me it's just fine if you implement a new method inside the Renderer class which does not touch the `draw()` calls or any other API more than what we agreed above ( see stencil enum topic ). But if you think you can "use" the renderer class from the outside, feel free to do so, up to you! If you need to call bgfx APIs though, please consider using the Renderer class instead, I don't want bgfx code to be spreaded all over the files.

> Other than the lighting class files described above, I was thinking to separate the walkmesh extraction/extrusion functions into separate files (currently in field.h/cpp). Other than that I don't see what else would require new files. I would appreciate if you specify what else you want me to separate.

Fine by me, if that has nothing to do with bgfx APIs, feel free to add that logic in separate files! It would make it also easier for you in the future to fire more PRs as that will most probably be your own pillar :)

> Thank you again for your time reviewing this PR, it is very much appreciated!

Again thank you! Looking forward for this bunch of changes :)

Reply:
> Thank you for understanding the need to reserve view id 0 for shadowmap rendering.

Thank you for taking the patience for still working on this. Really, really appreciate it!

> May I ask what made you change your mind on this? Also does this refer only to the shadowmap rendering shader or also the shader that rendered the walkmesh for the field shadows?

As we do reserve an entire view, it makes sense to use also separate shaders. As you will probably fire your own draw calls or whatever needed, you can totally bind your own shaders when calling them, and leave the game shaders to draw as usual. This makes this PR safer to merge as "if I enable the flag, I get something done in the view 0, if not, I still get what I used to have, but starting from view 1". That's basically a way to "kill" and entire block of code with a piece of mind when merging :)

> If I understand correctly, this means reverting to a similar state as when I first submitted this PR, where a separate drawShadowMap() function was used, is that correct? Also may I ask why this change of mind?

See the answer above. That applies also for this case. Related to the previous state, yes, feel free to encapsulate all your code in a separate function, or better, see the answers below :)

> Bind the shadowmap texture. This could be done without touching the draw function with the Renderer::useTexture() function, is that correct?

No, that function is to bind texture attachments when drawing something. Not sure how the light texture works but if that makes your day, feel free to use it ( as it is! ) in your own function code :)

> Set the stencil buffer. There is not currently a way to set the stencil buffer so I was thinking to add some APIs to set the stencil function and reference value from outside the draw() function. However, the actual setting by calling the bgfx functions would be inside the draw() function in similar to how the blend state is set. Would that be ok?

Absolutely fine to have the Stencil part exactly like we have the blend state for eg. Feel free to implement an enum, a relative function to set which state you want and let the `draw()` API set it based on which is the current enum set. Remember to reset its state in the `reset()` function. That's fine by me!

> You say here to not change anything from the core shaders, but lighting calculations as well as fetching the shadowmap needs to be done here so I'm not sure I understand your request (most of it is already is in include files anyway).

Whatever calculation you need to do, consider doing it in your own new shader vert and frag files. Use those ones as your own greenfield and remember to attach them when drawing :)

> If I understand you correctly, this means reverting to something similar as when I was using separate draw functions such as drawShadowMap() but instead of implementing them inside the Renderer class, I create a different class and source files (lighting.h/cpp) that contains the lighting related draw functions as well as the code found in common.cpp. Is this correct?

Well, that's up to you if it makes it easier. For me it's just fine if you implement a new method inside the Renderer class which does not touch the `draw()` calls or any other API more than what we agreed above ( see stencil enum topic ). But if you think you can "use" the renderer class from the outside, feel free to do so, up to you! If you need to call bgfx APIs though, please consider using the Renderer class instead, I don't want bgfx code to be spreaded all over the files.

> Other than the lighting class files described above, I was thinking to separate the walkmesh extraction/extrusion functions into separate files (currently in field.h/cpp). Other than that I don't see what else would require new files. I would appreciate if you specify what else you want me to separate.

Fine by me, if that has nothing to do with bgfx APIs, feel free to add that logic in separate files! It would make it also easier for you in the future to fire more PRs as that will most probably be your own pillar :)

> Thank you again for your time reviewing this PR, it is very much appreciated!

Again thank you! Looking forward for this bunch of changes :)

Reply:
I made some more changes as requested above.

- All lighting calculations are now done on separate new vert/frag shader files.
- Most of the lighting/shadow implementation has been moved to its own source files (lighting.h/cpp).
- Changes to the Renderer class have been reduced as much as possible.
- Added APIs to set the stencil buffer operations

Also, thankfully I was finally able to purchase a copy of FF8. I tested it with the changes of this PR and everything seems to work just fine. Let me know what you think!